File: Go (Golang)/fragment.go
// УСТАНОВКА ЗАВИСИМОСТЕЙ:
// go get github.com/xssnick/tonutils-go
// go get github.com/xssnick/tonutils-go/ton
// go get github.com/xssnick/tonutils-go/ton/wallet

package main

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/xssnick/tonutils-go/address"
	"github.com/xssnick/tonutils-go/liteclient"
	"github.com/xssnick/tonutils-go/ton"
	"github.com/xssnick/tonutils-go/ton/wallet"
)

// КОНФИГУРАЦИЯ
var (
	MNEMONIC = []string{
		"penalty", "undo", "fame", "place", "brand", "south", "lunar", "cage",
		"coconut", "girl", "lyrics", "ozone", "fence", "riot", "apology", "diagram",
		"nature", "manage", "there", "brief", "wet", "pole", "debris", "annual",
	}

	DATA = map[string]string{
		"stel_ssid":      "ваш_ssid",
		"stel_dt":        "-240",
		"stel_ton_token": "ваш_ton_token",
		"stel_token":     "ваш_token",
	}

	FRAGMENT_HASH      = "ed3ec875a724358cea"
	FRAGMENT_PUBLICKEY = "91b296c356bb0894b40397b54565c11f4b29ea610b8e14d2ae1136a50c5d1d03"
	FRAGMENT_WALLETS   = "te6cckECFgEAArEAAgE0AQsBFP8A9KQT9LzyyAsCAgEgAwYCAUgMBAIBIAgFABm+Xw9qJoQICg65D6AsAQLyBwEeINcLH4IQc2lnbrry4Ip/DQIBIAkTAgFuChIAGa3OdqJoQCDrkOuF/8AAUYAAAAA///+Il7w6CtQZIMze2+aVZS87QjJHoU5yqUljL1aSwzvDrCugAtzQINdJwSCRW49jINcLHyCCEGV4dG69IYIQc2ludL2wkl8D4IIQZXh0brqOtIAg1yEB0HTXIfpAMPpE+Cj6RDBYvZFb4O1E0IEBQdch9AWDB/QOb6ExkTDhgEDXIXB/2zzgMSDXSYECgLmRMOBw4g4NAeaO8O2i7fshgwjXIgKDCNcjIIAg1yHTH9Mf0x/tRNDSANMfINMf0//XCgAK+QFAzPkQmiiUXwrbMeHywIffArNQB7Dy0IRRJbry4IVQNrry4Ib4I7vy0IgikvgA3gGkf8jKAMsfAc8Wye1UIJL4D95w2zzYDgP27aLt+wL0BCFukmwhjkwCIdc5MHCUIccAs44tAdcoIHYeQ2wg10nACPLgkyDXSsAC8uCTINcdBscSwgBSMLDy0InXTNc5MAGk6GwShAe78uCT10rAAPLgk+1V4tIAAcAAkVvg69csCBQgkXCWAdcsCBwS4lIQseMPINdKERAPABCTW9sx4ddM0AByMNcsCCSOLSHy4JLSAO1E0NIAURO68tCPVFAwkTGcAYEBQNch1woA8uCO4sjKAFjPFsntVJPywI3iAJYB+kAB+kT4KPpEMFi68uCR7UTQgQFB1xj0BQSdf8jKAEAEgwf0U/Lgi44UA4MH9Fvy4Iwi1woAIW4Bs7Dy0JDiyFADzxYS9ADJ7VQAGa8d9qJoQBDrkOuFj8ACAUgVFAARsmL7UTQ1woAgABezJftRNBx1yHXCx+B27MAq"
	FRAGMENT_ADDRESS   = "0:20c429e3bb195f46a582c10eb687c6ed182ec58237a55787f245ec992c337118"
)

// ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ
func getCookies(data map[string]string) map[string]string {
	return map[string]string{
		"stel_ssid":      data["stel_ssid"],
		"stel_dt":        data["stel_dt"],
		"stel_ton_token": data["stel_ton_token"],
		"stel_token":     data["stel_token"],
	}
}

func fixBase64Padding(b64String string) string {
	missingPadding := len(b64String) % 4
	if missingPadding > 0 {
		b64String += strings.Repeat("=", 4-missingPadding)
	}
	return b64String
}

func cookiesToString(cookies map[string]string) string {
	var parts []string
	for key, value := range cookies {
		parts = append(parts, fmt.Sprintf("%s=%s", key, value))
	}
	return strings.Join(parts, "; ")
}

// FRAGMENT CLIENT
type FragmentClient struct {
	url     string
	cookies map[string]string
	client  *http.Client
}

func NewFragmentClient(fragmentHash string, cookiesData map[string]string) *FragmentClient {
	return &FragmentClient{
		url:     fmt.Sprintf("https://fragment.com/api?hash=%s", fragmentHash),
		cookies: getCookies(cookiesData),
		client:  &http.Client{Timeout: 30 * time.Second},
	}
}

func (fc *FragmentClient) FetchRecipient(query string) (string, error) {
	data := url.Values{}
	data.Set("query", query)
	data.Set("method", "searchStarsRecipient")

	req, err := http.NewRequest("POST", fc.url, strings.NewReader(data.Encode()))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Cookie", cookiesToString(fc.cookies))

	resp, err := fc.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", err
	}

	fmt.Println("Recipient search:", string(body))

	if found, ok := result["found"].(map[string]interface{}); ok {
		if recipient, ok := found["recipient"].(string); ok {
			return recipient, nil
		}
	}

	return "", fmt.Errorf("получатель не найден")
}

func (fc *FragmentClient) FetchReqId(recipient string, quantity int) (string, error) {
	data := url.Values{}
	data.Set("recipient", recipient)
	data.Set("quantity", strconv.Itoa(quantity))
	data.Set("method", "initBuyStarsRequest")

	req, err := http.NewRequest("POST", fc.url, strings.NewReader(data.Encode()))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Cookie", cookiesToString(fc.cookies))

	resp, err := fc.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", err
	}

	fmt.Println("Request ID:", string(body))

	if reqId, ok := result["req_id"].(string); ok {
		return reqId, nil
	}

	return "", fmt.Errorf("не удалось создать запрос")
}

func (fc *FragmentClient) FetchBuyLink(recipient, reqId string, quantity int) (string, string, string, error) {
	features, _ := json.Marshal([]interface{}{
		"SendTransaction",
		map[string]interface{}{"name": "SendTransaction", "maxMessages": 255},
	})

	data := url.Values{}
	data.Set("address", FRAGMENT_ADDRESS)
	data.Set("chain", "-239")
	data.Set("walletStateInit", FRAGMENT_WALLETS)
	data.Set("publicKey", FRAGMENT_PUBLICKEY)
	data.Set("features", string(features))
	data.Set("maxProtocolVersion", "2")
	data.Set("platform", "iphone")
	data.Set("appName", "Tonkeeper")
	data.Set("appVersion", "5.0.14")
	data.Set("transaction", "1")
	data.Set("id", reqId)
	data.Set("show_sender", "0")
	data.Set("method", "getBuyStarsLink")

	req, err := http.NewRequest("POST", fc.url, strings.NewReader(data.Encode()))
	if err != nil {
		return "", "", "", err
	}

	req.Header.Set("Accept", "application/json, text/javascript, */*; q=0.01")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")
	req.Header.Set("Origin", "https://fragment.com")
	req.Header.Set("Referer", fmt.Sprintf("https://fragment.com/stars/buy?recipient=%s&quantity=%d", recipient, quantity))
	req.Header.Set("User-Agent", "Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15")
	req.Header.Set("X-Requested-With", "XMLHttpRequest")
	req.Header.Set("Cookie", cookiesToString(fc.cookies))

	resp, err := fc.client.Do(req)
	if err != nil {
		return "", "", "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", "", "", err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", "", "", err
	}

	fmt.Println("Buy link:", string(body))

	if ok, _ := result["ok"].(bool); ok {
		if transaction, ok := result["transaction"].(map[string]interface{}); ok {
			if messages, ok := transaction["messages"].([]interface{}); ok && len(messages) > 0 {
				if msg, ok := messages[0].(map[string]interface{}); ok {
					addr := msg["address"].(string)
					amount := msg["amount"].(string)
					payload := msg["payload"].(string)
					return addr, amount, payload, nil
				}
			}
		}
	}

	return "", "", "", fmt.Errorf("не удалось получить данные транзакции")
}

// TON TRANSACTION
type TonTransaction struct {
	mnemonic []string
	client   *ton.APIClient
}

func NewTonTransaction(mnemonic []string) (*TonTransaction, error) {
	// Подключаемся к TON через публичный API
	client := liteclient.NewConnectionPool()
	
	// Используем публичный config
	err := client.AddConnectionsFromConfigUrl(context.Background(), "https://ton.org/global-config.json")
	if err != nil {
		return nil, fmt.Errorf("ошибка подключения к TON: %w", err)
	}

	apiClient := ton.NewAPIClient(client)

	return &TonTransaction{
		mnemonic: mnemonic,
		client:   apiClient,
	}, nil
}

func (tt *TonTransaction) DecodePayload(payloadBase64 string, starsCount int) string {
	fixed := fixBase64Padding(payloadBase64)
	decoded, err := base64.StdEncoding.DecodeString(fixed)
	if err != nil {
		return payloadBase64
	}

	var decodedText strings.Builder
	for _, b := range decoded {
		if b >= 32 && b < 127 {
			decodedText.WriteByte(b)
		} else {
			decodedText.WriteByte(' ')
		}
	}

	cleanText := regexp.MustCompile(`\s+`).ReplaceAllString(decodedText.String(), " ")
	cleanText = strings.TrimSpace(cleanText)

	pattern := regexp.MustCompile(fmt.Sprintf(`%d Telegram Stars.*`, starsCount))
	if match := pattern.FindString(cleanText); match != "" {
		return match
	}

	return cleanText
}

func (tt *TonTransaction) SendTransaction(ctx context.Context, recipientAddress string, amountTon float64, payloadBase64 string, starsCount int) (string, error) {
	fmt.Println("\n🔐 Инициализация кошелька...")

	// Создаем кошелек из мнемоники
	w, err := wallet.FromSeed(ctx, tt.client, tt.mnemonic, wallet.V4R2)
	if err != nil {
		return "", fmt.Errorf("ошибка создания кошелька: %w", err)
	}

	walletAddress := w.Address().String()
	fmt.Printf("✅ Адрес кошелька: %s\n", walletAddress)

	// Получаем баланс
	block, err := tt.client.CurrentMasterchainInfo(ctx)
	if err != nil {
		return "", fmt.Errorf("ошибка получения блока: %w", err)
	}

	balance, err := w.GetBalance(ctx, block)
	if err != nil {
		return "", fmt.Errorf("ошибка получения баланса: %w", err)
	}

	fmt.Printf("💰 Баланс: %s TON\n", balance.String())

	// Конвертируем сумму в nanoTON
	amountNano := uint64(amountTon * 1e9)

	// Декодируем payload
	payloadDecoded := tt.DecodePayload(payloadBase64, starsCount)

	fmt.Println("\n💸 Отправка транзакции...")
	fmt.Printf("   Получатель: %s\n", recipientAddress)
	fmt.Printf("   Сумма: %.4f TON (%d nanoTON)\n", amountTon, amountNano)
	fmt.Printf("   Комментарий: %s\n", payloadDecoded)

	// Парсим адрес получателя
	destAddr, err := address.ParseAddr(recipientAddress)
	if err != nil {
		return "", fmt.Errorf("неверный адрес получателя: %w", err)
	}

	// Декодируем payload из Base64 для body
	payloadBytes, err := base64.StdEncoding.DecodeString(fixBase64Padding(payloadBase64))
	if err != nil {
		return "", fmt.Errorf("ошибка декодирования payload: %w", err)
	}

	// Отправляем транзакцию
	err = w.Send(ctx, &wallet.Message{
		Mode: 3,
		InternalMessage: &wallet.InternalMessage{
			Bounce:  true,
			DstAddr: destAddr,
			Amount:  amountNano,
			Body:    payloadBytes,
		},
	}, true)

	if err != nil {
		return "", fmt.Errorf("ошибка отправки транзакции: %w", err)
	}

	// Генерируем хеш транзакции (упрощенно)
	txHash := fmt.Sprintf("%x", payloadBytes[:32])

	fmt.Println("\n✅ Транзакция отправлена успешно!")
	fmt.Printf("📝 Hash: %s\n", txHash)

	return txHash, nil
}

func (tt *TonTransaction) GetBalance(ctx context.Context) (string, error) {
	w, err := wallet.FromSeed(ctx, tt.client, tt.mnemonic, wallet.V4R2)
	if err != nil {
		return "", err
	}

	block, err := tt.client.CurrentMasterchainInfo(ctx)
	if err != nil {
		return "", err
	}

	balance, err := w.GetBalance(ctx, block)
	if err != nil {
		return "", err
	}

	return balance.String(), nil
}

// ОСНОВНОЙ ПРОЦЕСС
func BuyStars(ctx context.Context, username string, starsCount int, fragmentHash string, cookiesData map[string]string, mnemonic []string) (bool, string, error) {
	fragment := NewFragmentClient(fragmentHash, cookiesData)
	ton, err := NewTonTransaction(mnemonic)
	if err != nil {
		return false, "", err
	}

	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("🌟 ПОКУПКА TELEGRAM STARS")
	fmt.Println(strings.Repeat("=", 60))

	// Проверяем баланс
	balance, _ := ton.GetBalance(ctx)
	fmt.Printf("💰 Баланс кошелька: %s TON\n", balance)

	// Шаг 1: Поиск получателя
	fmt.Printf("\n📍 Шаг 1: Поиск получателя %s...\n", username)
	recipient, err := fragment.FetchRecipient(username)
	if err != nil {
		fmt.Println("❌ Получатель не найден")
		return false, "", err
	}
	fmt.Printf("✅ Получатель найден: %s\n", recipient)

	// Шаг 2: Создание запроса
	fmt.Printf("\n📝 Шаг 2: Создание запроса на %d звезд...\n", starsCount)
	reqId, err := fragment.FetchReqId(recipient, starsCount)
	if err != nil {
		fmt.Println("❌ Не удалось создать запрос")
		return false, "", err
	}
	fmt.Printf("✅ Request ID: %s\n", reqId)

	// Шаг 3: Получение данных транзакции
	fmt.Println("\n🔍 Шаг 3: Получение данных транзакции...")
	addr, amount, payload, err := fragment.FetchBuyLink(recipient, reqId, starsCount)
	if err != nil {
		fmt.Println("❌ Не удалось получить данные транзакции")
		return false, "", err
	}

	amountInt, _ := strconv.ParseInt(amount, 10, 64)
	amountTon := float64(amountInt) / 1e9

	fmt.Printf("✅ Сумма к оплате: %.4f TON\n", amountTon)
	fmt.Printf("✅ Адрес Fragment: %s\n", addr)

	// Шаг 4: Отправка TON
	fmt.Println("\n💳 Шаг 4: Отправка транзакции в блокчейн...")
	txHash, err := ton.SendTransaction(ctx, addr, amountTon, payload, starsCount)
	if err != nil {
		fmt.Printf("❌ Ошибка при отправке: %v\n", err)
		return false, "", err
	}

	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("🎉 ПОКУПКА ЗАВЕРШЕНА УСПЕШНО!")
	fmt.Println(strings.Repeat("=", 60))

	return true, txHash, nil
}

func main() {
	ctx := context.Background()

	// Параметры покупки
	username := "@example"  // Замените на реальный username
	starsCount := 100

	success, txHash, err := BuyStars(
		ctx,
		username,
		starsCount,
		FRAGMENT_HASH,
		DATA,
		MNEMONIC,
	)

	if err != nil {
		fmt.Printf("\n💥 Критическая ошибка: %v\n", err)
		return
	}

	if success {
		fmt.Println("\n🔗 Просмотр транзакции:")
		fmt.Printf("   https://tonviewer.com/transaction/%s\n", txHash)
		fmt.Printf("   https://tonscan.org/tx/%s\n", txHash)
	} else {
		fmt.Println("\n❌ Покупка не удалась. Проверьте конфигурацию.")
	}
}

File: TypeScript/fragment.ts
// УСТАНОВКА ЗАВИСИМОСТЕЙ:
// npm install tonweb tonweb-mnemonic node-fetch @types/node

import TonWeb from 'tonweb';
import * as tonMnemonic from 'tonweb-mnemonic';
import fetch from 'node-fetch';

// ТИПЫ
interface CookiesData {
    stel_ssid: string;
    stel_dt: string;
    stel_ton_token: string;
    stel_token: string;
}

interface Cookies {
    [key: string]: string;
}

interface RecipientResponse {
    found?: {
        recipient: string;
    };
}

interface ReqIdResponse {
    req_id?: string;
}

interface BuyLinkResponse {
    ok?: boolean;
    transaction?: {
        messages: Array<{
            address: string;
            amount: string;
            payload: string;
        }>;
    };
}

interface BuyStarsResult {
    success: boolean;
    txHash: string | null;
}

// КОНФИГУРАЦИЯ
const API_TON: string = "ваш_api_ключ";

const MNEMONIC: string[] = [
    "penalty", "undo", "fame", "place", "brand", "south", "lunar", "cage",
    "coconut", "girl", "lyrics", "ozone", "fence", "riot", "apology", "diagram",
    "nature", "manage", "there", "brief", "wet", "pole", "debris", "annual"
];

const DATA: CookiesData = {
    stel_ssid: "ваш_ssid",
    stel_dt: "-240",
    stel_ton_token: "ваш_ton_token",
    stel_token: "ваш_token"
};

const FRAGMENT_HASH: string = "ed3ec875a724358cea";
const FRAGMENT_PUBLICKEY: string = "91b296c356bb0894b40397b54565c11f4b29ea610b8e14d2ae1136a50c5d1d03";
const FRAGMENT_WALLETS: string = "te6cckECFgEAArEAAgE0AQsBFP8A9KQT9LzyyAsCAgEgAwYCAUgMBAIBIAgFABm+Xw9qJoQICg65D6AsAQLyBwEeINcLH4IQc2lnbrry4Ip/DQIBIAkTAgFuChIAGa3OdqJoQCDrkOuF/8AAUYAAAAA///+Il7w6CtQZIMze2+aVZS87QjJHoU5yqUljL1aSwzvDrCugAtzQINdJwSCRW49jINcLHyCCEGV4dG69IYIQc2ludL2wkl8D4IIQZXh0brqOtIAg1yEB0HTXIfpAMPpE+Cj6RDBYvZFb4O1E0IEBQdch9AWDB/QOb6ExkTDhgEDXIXB/2zzgMSDXSYECgLmRMOBw4g4NAeaO8O2i7fshgwjXIgKDCNcjIIAg1yHTH9Mf0x/tRNDSANMfINMf0//XCgAK+QFAzPkQmiiUXwrbMeHywIffArNQB7Dy0IRRJbry4IVQNrry4Ib4I7vy0IgikvgA3gGkf8jKAMsfAc8Wye1UIJL4D95w2zzYDgP27aLt+wL0BCFukmwhjkwCIdc5MHCUIccAs44tAdcoIHYeQ2wg10nACPLgkyDXSsAC8uCTINcdBscSwgBSMLDy0InXTNc5MAGk6GwShAe78uCT10rAAPLgk+1V4tIAAcAAkVvg69csCBQgkXCWAdcsCBwS4lIQseMPINdKERAPABCTW9sx4ddM0AByMNcsCCSOLSHy4JLSAO1E0NIAURO68tCPVFAwkTGcAYEBQNch1woA8uCO4sjKAFjPFsntVJPywI3iAJYB+kAB+kT4KPpEMFi68uCR7UTQgQFB1xj0BQSdf8jKAEAEgwf0U/Lgi44UA4MH9Fvy4Iwi1woAIW4Bs7Dy0JDiyFADzxYS9ADJ7VQAGa8d9qJoQBDrkOuFj8ACAUgVFAARsmL7UTQ1woAgABezJftRNBx1yHXCx+B27MAq";
const FRAGMENT_ADDRESS: string = "0:20c429e3bb195f46a582c10eb687c6ed182ec58237a55787f245ec992c337118";

// ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ
function getCookies(data: CookiesData): Cookies {
    return {
        stel_ssid: data.stel_ssid || "",
        stel_dt: data.stel_dt || "",
        stel_ton_token: data.stel_ton_token || "",
        stel_token: data.stel_token || ""
    };
}

function fixBase64Padding(b64String: string): string {
    const missingPadding = b64String.length % 4;
    if (missingPadding) {
        b64String += "=".repeat(4 - missingPadding);
    }
    return b64String;
}

function cookiesToString(cookies: Cookies): string {
    return Object.entries(cookies)
        .map(([key, value]) => `${key}=${value}`)
        .join("; ");
}

// FRAGMENT CLIENT
class FragmentClient {
    private url: string;
    private cookies: Cookies;

    constructor(fragmentHash: string, cookiesData: CookiesData) {
        this.url = `https://fragment.com/api?hash=${fragmentHash}`;
        this.cookies = getCookies(cookiesData);
    }

    async fetchRecipient(query: string): Promise<string | null> {
        const data = new URLSearchParams({
            query: query,
            method: "searchStarsRecipient"
        });

        const response = await fetch(this.url, {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "Cookie": cookiesToString(this.cookies)
            },
            body: data.toString()
        });

        const result: RecipientResponse = await response.json() as RecipientResponse;
        console.log("Recipient search:", result);
        return result.found?.recipient || null;
    }

    async fetchReqId(recipient: string, quantity: number): Promise<string | null> {
        const data = new URLSearchParams({
            recipient: recipient,
            quantity: quantity.toString(),
            method: "initBuyStarsRequest"
        });

        const response = await fetch(this.url, {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "Cookie": cookiesToString(this.cookies)
            },
            body: data.toString()
        });

        const result: ReqIdResponse = await response.json() as ReqIdResponse;
        console.log("Request ID:", result);
        return result.req_id || null;
    }

    async fetchBuyLink(
        recipient: string,
        reqId: string,
        quantity: number
    ): Promise<{ address: string | null; amount: string | null; payload: string | null }> {
        const data = new URLSearchParams({
            address: FRAGMENT_ADDRESS,
            chain: "-239",
            walletStateInit: FRAGMENT_WALLETS,
            publicKey: FRAGMENT_PUBLICKEY,
            features: JSON.stringify(["SendTransaction", { name: "SendTransaction", maxMessages: 255 }]),
            maxProtocolVersion: "2",
            platform: "iphone",
            appName: "Tonkeeper",
            appVersion: "5.0.14",
            transaction: "1",
            id: reqId,
            show_sender: "0",
            method: "getBuyStarsLink"
        });

        const response = await fetch(this.url, {
            method: "POST",
            headers: {
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
                "Origin": "https://fragment.com",
                "Referer": `https://fragment.com/stars/buy?recipient=${recipient}&quantity=${quantity}`,
                "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15",
                "X-Requested-With": "XMLHttpRequest",
                "Cookie": cookiesToString(this.cookies)
            },
            body: data.toString()
        });

        const jsonData: BuyLinkResponse = await response.json() as BuyLinkResponse;
        console.log("Buy link:", jsonData);

        if (jsonData.ok && jsonData.transaction && jsonData.transaction.messages.length > 0) {
            const msg = jsonData.transaction.messages[0];
            return {
                address: msg.address,
                amount: msg.amount,
                payload: msg.payload
            };
        }

        return { address: null, amount: null, payload: null };
    }
}

// TON TRANSACTION
class TonTransaction {
    private mnemonic: string[];
    private tonweb: typeof TonWeb;

    constructor(mnemonic: string[]) {
        this.mnemonic = mnemonic;
        this.tonweb = new TonWeb(new TonWeb.HttpProvider('https://toncenter.com/api/v2/jsonRPC'));
    }

    decodePayload(payloadBase64: string, starsCount: number): string {
        try {
            const fixedBase64 = fixBase64Padding(payloadBase64);
            const decodedBytes = Buffer.from(fixedBase64, 'base64');

            let decodedText = "";
            for (let i = 0; i < decodedBytes.length; i++) {
                const byte = decodedBytes[i];
                decodedText += (byte >= 32 && byte < 127) ? String.fromCharCode(byte) : " ";
            }

            const cleanText = decodedText.replace(/\s+/g, " ").trim();

            const regex = new RegExp(`${starsCount} Telegram Stars.*`);
            const match = cleanText.match(regex);
            return match ? match[0] : cleanText;
        } catch (error) {
            console.error("Ошибка декодирования payload:", error);
            return payloadBase64;
        }
    }

    async sendTransaction(
        recipientAddress: string,
        amountNano: number,
        payload: string,
        starsCount: number
    ): Promise<string> {
        try {
            console.log("\n🔐 Инициализация кошелька...");

            // Получаем ключи из мнемоники
            const mnemonicArray = this.mnemonic;
            const keyPair = await tonMnemonic.mnemonicToKeyPair(mnemonicArray);

            // Создаем кошелек V4R2
            const WalletClass = this.tonweb.wallet.all.v4R2;
            const wallet = new WalletClass(this.tonweb.provider, {
                publicKey: keyPair.publicKey,
                wc: 0
            });

            const walletAddress = await wallet.getAddress();
            console.log(`✅ Адрес кошелька: ${walletAddress.toString(true, true, true)}`);

            // Получаем seqno
            const seqno: number = (await wallet.methods.seqno().call()) || 0;
            console.log(`📊 Seqno: ${seqno}`);

            // Декодируем payload из Base64 в BOC
            const payloadCell = TonWeb.boc.Cell.oneFromBoc(
                TonWeb.utils.base64ToBytes(fixBase64Padding(payload))
            );

            // Конвертируем сумму в nanotons
            const amountInNano = this.tonweb.utils.toNano(amountNano.toString());

            console.log(`\n💸 Отправка транзакции...`);
            console.log(`   Получатель: ${recipientAddress}`);
            console.log(`   Сумма: ${amountNano} TON (${amountInNano} nanoTON)`);
            console.log(`   Комментарий: ${this.decodePayload(payload, starsCount)}`);

            // Создаем и отправляем транзакцию
            const transfer = wallet.methods.transfer({
                secretKey: keyPair.secretKey,
                toAddress: recipientAddress,
                amount: amountInNano,
                seqno: seqno,
                payload: payloadCell,
                sendMode: 3
            });

            await transfer.send();

            // Получаем hash транзакции
            const txHash = TonWeb.utils.bytesToBase64(await transfer.getQuery().hash());

            console.log(`\n✅ Транзакция отправлена успешно!`);
            console.log(`📝 Hash: ${txHash}`);

            return txHash;

        } catch (error) {
            console.error("\n❌ Ошибка при отправке транзакции:", error);
            throw error;
        }
    }

    async getBalance(): Promise<string> {
        try {
            const mnemonicArray = this.mnemonic;
            const keyPair = await tonMnemonic.mnemonicToKeyPair(mnemonicArray);

            const WalletClass = this.tonweb.wallet.all.v4R2;
            const wallet = new WalletClass(this.tonweb.provider, {
                publicKey: keyPair.publicKey,
                wc: 0
            });

            const walletAddress = await wallet.getAddress();
            const balance = await this.tonweb.getBalance(walletAddress);
            const balanceInTon = this.tonweb.utils.fromNano(balance);

            console.log(`💰 Баланс кошелька: ${balanceInTon} TON`);
            return balanceInTon;

        } catch (error) {
            console.error("Ошибка получения баланса:", error);
            return "0";
        }
    }
}

// ОСНОВНОЙ ПРОЦЕСС ПОКУПКИ
async function buyStars(
    username: string,
    starsCount: number,
    fragmentHash: string,
    cookiesData: CookiesData,
    mnemonic: string[]
): Promise<BuyStarsResult> {
    const fragment = new FragmentClient(fragmentHash, cookiesData);
    const ton = new TonTransaction(mnemonic);

    console.log("\n" + "=".repeat(60));
    console.log("🌟 ПОКУПКА TELEGRAM STARS");
    console.log("=".repeat(60));

    // Проверяем баланс
    await ton.getBalance();

    // Шаг 1: Поиск получателя
    console.log(`\n📍 Шаг 1: Поиск получателя ${username}...`);
    const recipient = await fragment.fetchRecipient(username);
    if (!recipient) {
        console.log("❌ Получатель не найден");
        return { success: false, txHash: null };
    }
    console.log(`✅ Получатель найден: ${recipient}`);

    // Шаг 2: Создание запроса
    console.log(`\n📝 Шаг 2: Создание запроса на ${starsCount} звезд...`);
    const reqId = await fragment.fetchReqId(recipient, starsCount);
    if (!reqId) {
        console.log("❌ Не удалось создать запрос");
        return { success: false, txHash: null };
    }
    console.log(`✅ Request ID: ${reqId}`);

    // Шаг 3: Получение данных транзакции
    console.log(`\n🔍 Шаг 3: Получение данных транзакции...`);
    const { address, amount, payload } = await fragment.fetchBuyLink(recipient, reqId, starsCount);
    if (!address || !amount || !payload) {
        console.log("❌ Не удалось получить данные транзакции");
        return { success: false, txHash: null };
    }

    const amountTon = parseFloat(amount) / 1_000_000_000;
    console.log(`✅ Сумма к оплате: ${amountTon.toFixed(4)} TON`);
    console.log(`✅ Адрес Fragment: ${address}`);

    // Шаг 4: Отправка TON
    console.log(`\n💳 Шаг 4: Отправка транзакции в блокчейн...`);
    try {
        const txHash = await ton.sendTransaction(address, amountTon, payload, starsCount);

        if (txHash) {
            console.log("\n" + "=".repeat(60));
            console.log("🎉 ПОКУПКА ЗАВЕРШЕНА УСПЕШНО!");
            console.log("=".repeat(60));
            return { success: true, txHash };
        }
    } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        console.error(`\n❌ Ошибка при отправке: ${errorMessage}`);
        return { success: false, txHash: null };
    }

    return { success: false, txHash: null };
}

// ПРИМЕР ИСПОЛЬЗОВАНИЯ
async function main(): Promise<void> {
    try {
        // Параметры покупки
        const username: string = "@example";  // Замените на реальный username
        const starsCount: number = 100;       // Количество звезд

        const { success, txHash } = await buyStars(
            username,
            starsCount,
            FRAGMENT_HASH,
            DATA,
            MNEMONIC
        );

        if (success && txHash) {
            console.log(`\n🔗 Просмотр транзакции:`);
            console.log(`   https://tonviewer.com/transaction/${txHash}`);
            console.log(`   https://tonscan.org/tx/${txHash}`);
        } else {
            console.log("\n❌ Покупка не удалась. Проверьте конфигурацию.");
        }

    } catch (error) {
        console.error("\n💥 Критическая ошибка:", error);
    }
}

// Запуск
if (require.main === module) {
    main().catch(console.error);
}

// Экспорт
export {
    FragmentClient,
    TonTransaction,
    buyStars,
    getCookies,
    fixBase64Padding,
    type CookiesData,
    type BuyStarsResult
};
